File: mobile/seeker-solana/src/services/TaskExecutor.ts
/**
 * TaskExecutor — Local heuristic-based task execution.
 *
 * For MVP, this uses simple text processing instead of actual AI models.
 * The three task types match what the coordinator sends:
 * - summarize: truncate to first ~200 chars + word count
 * - classify: keyword-based sentiment analysis
 * - extract_json: regex extraction of JSON blocks from text
 *
 * This mirrors the desktop worker's taskExecutor.ts exactly,
 * so coordinators treat mobile and desktop results the same way.
 */
import type { TaskType } from "./protocol";

/**
 * Execute a TASK job locally.
 * Returns structured output matching the coordinator's expected format.
 */
export function executeTask(taskType: TaskType, input: string): unknown {
  switch (taskType) {
    case "summarize":
      return summarize(input);
    case "classify":
      return classify(input);
    case "extract_json":
      return extractJson(input);
    default:
      return { error: `Unknown task_type: ${taskType as string}` };
  }
}

// ── Summarize ──────────────────────────────────

function summarize(input: string): { summary: string; word_count: number } {
  const words = input.trim().split(/\s+/);
  const truncated =
    input.length > 200 ? input.slice(0, 200) + "..." : input;
  return { summary: truncated, word_count: words.length };
}

// ── Classify ───────────────────────────────────

function classify(
  input: string
): { sentiment: "positive" | "negative" | "neutral"; confidence: number } {
  const lower = input.toLowerCase();
  const positiveWords = [
    "good", "great", "excellent", "amazing", "love",
    "happy", "best", "wonderful", "fantastic",
  ];
  const negativeWords = [
    "bad", "terrible", "awful", "hate", "worst",
    "horrible", "poor", "disappointing", "sad",
  ];

  let score = 0;
  for (const w of positiveWords) if (lower.includes(w)) score++;
  for (const w of negativeWords) if (lower.includes(w)) score--;

  const sentiment =
    score > 0 ? "positive" : score < 0 ? "negative" : "neutral";
  const confidence = Math.min(1, Math.abs(score) * 0.2 + 0.3);
  return { sentiment, confidence };
}

// ── Extract JSON ───────────────────────────────

function extractJson(
  input: string
): { extracted: unknown[]; count: number } {
  const jsonPattern = /\{[^{}]*(?:\{[^{}]*\}[^{}]*)*\}/g;
  const matches = input.match(jsonPattern) ?? [];
  const parsed: unknown[] = [];

  for (const m of matches) {
    try {
      parsed.push(JSON.parse(m));
    } catch {
      // skip invalid JSON blocks
    }
  }

  return { extracted: parsed, count: parsed.length };
}

File: mobile/seeker-worker-android/app/src/main/java/com/openclaw/seeker/TaskExecutor.kt
package com.openclaw.seeker

import org.json.JSONArray
import org.json.JSONObject

/**
 * Executes TASK jobs locally on the device.
 */
object TaskExecutor {

    fun execute(taskType: String, input: String): JSONObject {
        return when (taskType) {
            "summarize" -> summarize(input)
            "classify" -> classify(input)
            "extract_json" -> extractJson(input)
            else -> JSONObject().put("error", "Unknown task_type: $taskType")
        }
    }

    private fun summarize(input: String): JSONObject {
        val words = input.trim().split("\\s+".toRegex())
        val truncated = if (input.length > 200) input.take(200) + "..." else input
        return JSONObject().apply {
            put("summary", truncated)
            put("word_count", words.size)
        }
    }

    private fun classify(input: String): JSONObject {
        val lower = input.lowercase()
        val positive = listOf("good", "great", "excellent", "amazing", "love", "happy")
        val negative = listOf("bad", "terrible", "awful", "hate", "worst", "horrible")

        var score = 0
        positive.forEach { if (lower.contains(it)) score++ }
        negative.forEach { if (lower.contains(it)) score-- }

        val sentiment = when {
            score > 0 -> "positive"
            score < 0 -> "negative"
            else -> "neutral"
        }
        return JSONObject().apply {
            put("sentiment", sentiment)
            put("confidence", minOf(1.0, Math.abs(score) * 0.2 + 0.3))
        }
    }

    private fun extractJson(input: String): JSONObject {
        val pattern = "\\{[^{}]*(?:\\{[^{}]*}[^{}]*)*}".toRegex()
        val matches = pattern.findAll(input)
        val extracted = JSONArray()

        for (match in matches) {
            try {
                extracted.put(JSONObject(match.value))
            } catch (e: Exception) {
                // skip invalid JSON
            }
        }

        return JSONObject().apply {
            put("extracted", extracted)
            put("count", extracted.length())
        }
    }
}
